File: _practice/1382-Balance-a-Binary-Search-Tree/practice.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        nodes = []
        def inorder(node):
            if not node: return
            inorder(node.left)
            nodes.append(node.val)
            inorder(node.right)
        inorder(root)

        def create_bst(l, r):
            if l > r: return None
            m = (l + r) // 2
            left = create_bst(l, m - 1)
            right = create_bst(m + 1, r)
            return TreeNode(nodes[m], left, right)

        return create_bst(0, len(nodes) - 1)    

File: _practice/1382-Balance-a-Binary-Search-Tree/practice.js
//  Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var balanceBST = function (root) {
  const nodes = [];
  const inorder = (node) => {
    if (!node) return;
    inorder(node.left);
    nodes.push(node.val);
    inorder(node.right);
  };
  inorder(root);

  const create_bst = (l, r) => {
    if (l > r) return null;
    const m = Math.floor((l + r) / 2);
    const left = create_bst(l, m - 1);
    const right = create_bst(m + 1, r);
    return new TreeNode(nodes[m], left, right);
  };

  return create_bst(0, nodes.length - 1);
};
